### ISSUE TITLE
Dependency on boltex fork

### ISSUE THREAD

**Issue by vic** [florinpatrascu/bolt_sips#5]
Is there still any reason to depend on `florinpatrascu/boltex` instead of `mschae/boltex` ? 

mschae/boltex#10 has been merged and would like to remove the TODO I had once we depend on the orignal boltex code.

**Comment by florinpatrascu** [comment]
Hi @vic - I forked Boltex for two main reasons: to have a stable release I could depend on and to allow Boltex itself to cover the remaining works - as illustrated by @mschae's todo list (see his [README](https://github.com/mschae/boltex/blob/master/README.md) w/o adding more pressure with my own specific functional requirements, as at that time the both drivers were developed almost in parallel :)

It is not my intention to maintain a bolted fork, of course. The only functionality we need, for us to use mschae/boltex directly, is the connectivity. As soon as the db_connection is merged, I will start refactoring bolt_sips and remove the pool_boy based implementation - I needed a connection pool, as you know.

I mean we can probably do the switch right now, but what it will happened when the DBConnection is merged into boltex?

**Comment by mschae** [comment]
Hey,

you are always welcome to use a fork although I found it bothersome to keep up to date.

You could work off of my dbconnection branch right now and see if it suits your needs. If not I'd be very happy to hear about it and keep improving it.
If it helps I can also start releasing 1.0.0alpha and beta releases.

Whatever is easiest for you, I definitely strive to keep the boltex branch as stable as possible and add all the functionality we need in the driver.

Thanks

**Comment by florinpatrascu** [comment]
Many thanks, @mschae! I'll try to dive into this later tonight, after work, and see how it's going. If the dbconnection works, and I am sure it will work, then I'll have to remove my poolboy code and delegate the connectivity to boltex, which ideally is the way it should be; have no interest in maintaining an external fork. I did that just as a safety buffer, until you got a chance to add a connection pool of some sort and stabilize the driver in general. I'll definitely keep you guys in sync.

**Comment by florinpatrascu** [comment]
Hi @mschae,

I started to refactor bolt_sips and I am using now your `add-dbconnection` branch; see  [bolt_sips/boltex_w_dbconn](https://github.com/florinpatrascu/bolt_sips/tree/boltex_w_dbconn),in my bolt_sips repo . Everything is working fine, but **without a pool** (when using de default: `DBConnection.Connection`)

However, when I am trying to delegate the pool to `DBConnection.Poolboy`, I am getting this error:

``` elixir
(UndefinedFunctionError) function :invalid_message.exception/1 is undefined 
  (module :invalid_message is not available)
```

So if you want to try, this is the Bolt.Sips config that will trigger the error:

``` elixir
config :bolt_sips, Bolt,
  host: "localhost",
  port: 7687,
  pool: DBConnection.Poolboy,
  pool_size: 5,
  pool_overflow: 1,
  timeout: 10_000
```

Do you have an working example using Boltex from your new branch, and the `DBConnection.Poolboy` implementation? Most probably I am doing something silly at my end...

Thank you

**Comment by florinpatrascu** [comment]
BTW @vic, this branch conforms to [mschae/boltex#10](https://github.com/mschae/boltex/pull/10) since it is using the boltex from @mschae's new branch, and the tests are  passing (except when using a connection pool, but that's another issue) 👍

**Comment by florinpatrascu** [comment]
@mschae - sorry, you don't have to check bolt_sips. I made a few minor changes to your `add-dbconnection` branch, please see this gist containing the patch [florinpatrascu/0001-attempt-to-use-Poolboy.patch](https://gist.github.com/florinpatrascu/f39ede4f6bd5b6e130301c9aad43c39d)   and I can see the same error when running the test from boltex.

**Comment by mschae** [comment]
Hey @florinpatrascu - sorry that error is really nondescript. I'll let the DBConnection maintainers know!

The way you are trying to integrate poolboy won't work. You need to check out connections you can run your queries on.

Try this:

``` Elixir
{:ok, pid} = DBConnection.start_link Boltex.Connection, [pool: DBConnection.Poolboy, host: "localhost", port: 7687, auth: {"neo4j", "password"}]

DBConnection.run pid, fn(worker) -> 
  DBConnection.execute(worker, %Boltex.Query{statement: "MATCH (n) RETURN n"}, %{}, []) 
end, [pool: DBConnection.Poolboy]
```

**Comment by florinpatrascu** [comment]
Thanks for the tip, @mschae! I am more familiar with using Poolboy directly, than through DBConn. Will try this, thanks again! So, Boltex will not use a pool? Not finding the pattern above anywhere in the Boltex code.

**Comment by mschae** [comment]
Not sure I understand... Where would boltex use a pool? Maybe I'll add tests but the beauty of DBConnection is that I merely write a driver, make it DBConnection compliant and they worry about properly implementing different pooling strategies, etc. Checkout Postgrex, all they do is pass the options down (which I might as well - but then what's the gain?).

The DBConnection branch is still experimental but I am assuming that such things as passing the pool options down will be done elsewhere.

Boltex is only intended as a low lever driver.

**Comment by florinpatrascu** [comment]
I thought you're planning to encapsulate the DBConnection and export its functionality via a Boltex api, therefore managing the pool  "behind the scenes" :) That way one will use only Boltex, not DBConnection directly, but I understand your point, thank you for clarifying this.

**Comment by mschae** [comment]
No worries. I don't want to wrap DBConnection - the user should choose what Pool they want to use etc.

I might provide connect and query functions analogous to Postgrex - not sure yet... :)

**Comment by florinpatrascu** [comment]
Makes sense :)

I'll keep my pool implementation around for awhile, and focus on integrating with Boltex as is. @mschae, do you have an ETA for merging the DBConnection branch and pushing a new Boltex build to Hex? Thanks again!

**Comment by florinpatrascu** [comment]
After spending a fair amount of time past weekend with integrating and debugging the merger of bolt_sips and Boltex-DBConn, I decided to use the original, dare to say: raw Boltex, by including its core in Bolt.Sips and manage the connectivity myself. I hope I won't regret this decision :(

This is because after I thought I stabilized the merge (work in the: boltex_w_dbconn branch) I started to see some very odd errors when executing simple or complex transactions. And since my plan was to work on the driver itself and on the interface with the client apps, going back to redesigning the connectivity and the integration with the new boltex version will consume all my time, sorry. 

@mschae - I'll monitor the Boltex encoding/decoding changes and integrate them back into my project when needed, and send any PR back when there are meaningful updates at our end. Later, when Boltex will have a first public release close to what you envision to be, I'll try again to rework my integration. Thanks again for Boltex and support.

**Comment by mschae** [comment]
@florinpatrascu Sorry to hear that the integration was not as straightforward as I'd hoped. If you could elaborate on those "very odd errors" - maybe even with stacktraces, queries and error messages - I could look into those.

Either way do note that the DBConnection implementation was to be considered alpha. I noted that multiple times.

Thanks

**Comment by florinpatrascu** [comment]
@mschae - no worries, I understood exactly the alpha status, but I thought I can manage the integration, eventually sending back some positive news. It didn't turn that way mostly because of my lack of understanding about DBConn internal functionality. No blame on you, don't worry :) With the little I understood, about DBConn, I realized that maybe I don't needed it for now, the pool I implemented is known (to me) and it can be easily debugged if need be; basic GenServer stuff, no fluff :) just kidding

Regarding the odd errors. I am calling them odd, because there are no Elixir errors per see, nor stacktraces. It's just the behavior I expected from tests is sometimes working and sometimes not: depending on load, pool occupancy, nodes addressed, etc. Mind you, it may be my fault completely, most probably that's the case, and a resulted screwed up pool management while using DBConn+DBConn.Poolboy. Basically I had transaction leaking outside their begin-end logic blocks, missing Data on queries, false Cypher deletes aka: I deleted a bunch of nodes, and they reappeared back seconds later but not every time .... moment when I gave up, since all of the above worked (even in production) with the "raw" packet enc/dec of Boltex and my own pool implementation. I also ran out of weekend time and I have a busy week ahead:/

I'll need to write more tests in order to catch these new issues and retry again the DBConn integration, but I also need a stable driver meanwhile, hence my approach last night. Thank you for your interest in this, @mschae!

**Comment by mschae** [comment]
Hey @florinpatrascu, thanks for elaborating. Any chance you can push up a branch with those intermittently failing tests so I can investigate?

I added an example on how to use the DBConnection - maybe it will help others trying to integrate Boltex:
https://github.com/mschae/boltex/tree/add-dbconnection/examples/pool_example

Thanks for your feedback!

**Comment by florinpatrascu** [comment]
Sure, I have a "most" stable version in this branch: https://github.com/florinpatrascu/bolt_sips/tree/boltex_w_dbconn. I'll check the examples later on, thank you for providing them!
